### PR TITLE
many-rs crates refactor

### LINKED ISSUE #35: Split `many-rs` into multiple crates
Right now there's 4:
- `many` itself
- `many-cli`
- `many-macros`
- `many-client`

We should split the `many` into more crates to separate the concerns;

- [ ] `many-server`, which would handle the transport and various traits regarding server creation and running.
- [ ] `many-types`, which contains types (like `identity`, `attribute`, etc).
- [ ] `many-modules`, which contains the module definitions
- [ ] `many` itself should contain the protocol related stuff (like creation of envelopes/messages)
- [ ] `many-hsm` to manage the identity HSM stuff

----

We should also take this time to remove/refactor/replace the whole `CoseKeyIdentity` type tree, which is convoluted and just plain hard to use/maintain. This can be done as a separate issue/step.

### PR DESCRIPTION
**Pull request by hansl**
Split all the crates into a specific list of crates that make more sense semantically and can also be individually imported without having to import the whole dependency graph.

There is a lot of follow up work to this, but at least I wanted to get this in so we can work with a baseline. Follow up should affect less lines of code.

The following was done as work that was not exactly just moving files around:

- [x] Rename `Identity` to `Address`. This was overdue.
- [x] Create a `Identity` trait that allows to sign messages. I should have done this as separate work, but this was simple enough. This will allow further split of `many-identity` crate into `many-identity-dsa` and `many-identity-hsm` which will make more sense.
- [x] The `many_module` macro had a bit of a refactoring and now includes the sender and return type as typed in the trait. That is because it was hard to keep clippy and the IDE happy both at the same time.

Fixes #35.

### CODE REVIEW

File: src/many/src/main.rs [hunk 1]
```diff
@@ -160,7 +160,7 @@ struct ServerOpt {
     addr: SocketAddr,
 
     /// The name to give the server.
-    #[clap(long, short, default_value = "many-server")]
+    #[clap(long, short, default_value = "many-server-server")]
```
Comment (fmorency) [inline src/many/src/main.rs:163]: That's a lot of servers.

**Review by fmorency — CHANGES_REQUESTED** [review]
First iteration. Thanks for doing this; I like where this is going.

- Would it be possible to create a diagram with the architecture, including crates and (rust) modules?
- Do you have a roadmap of the changes you plan to make?

File: src/many-identity/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+[dev-dependencies]
+many-identity = { path = ".", features = [ "serde", "testing" ] }
+proptest = "1.0.0"
+serde_test = "1.0.139"
+
+[features]
+default = ["cose", "minicbor"]
+cose = []
+hsm_test = []
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-identity/Cargo.toml:38]: Yes, but it will become way clearer in the follow up PR where `hsm` becomes its own crate so there's no need to hide everything behind a feature. As such I'll keep this as is for now.

File: src/many-identity/src/identity.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//! An Identity is a signer that also has an address on the MANY protocol.
+use crate::Address;
+use many_error::ManyError;
+
+pub(crate) mod cose;
+
+pub use cose::*;
+
+pub trait Identity {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-identity/src/identity.rs:9]: No. I thought about it, but it makes it complex as `verify` only requires the public key, but an Identity has both. We can always add it later.

Verification is really a function. We already have a `Verifier` trait from the `signature` crate which covers this for types.

File: src/many-identity/src/identity/cose.rs [hunk 4]
```diff
@@ -44,7 +44,7 @@ impl Signature for CoseKeyIdentitySignature {
 
 #[derive(Clone, Debug, PartialEq)]
 pub struct CoseKeyIdentity {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-identity/src/identity/cose.rs:46]: This will implement that trait. I'm planning on splitting `CoseKeyIdentity` into multiple types; `Ed25519Identity`, `EcDsaIdentity` and `HsmIdentity`. So in a sense, we will get rid of this schizophrenic type.

File: src/many-identity/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+[package]
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-identity/Cargo.toml:1]: Yes, separate PR. https://github.com/liftedinit/many-rs/issues/161

File: src/many-macros/src/lib.rs [hunk 6]
```diff
@@ -122,14 +118,23 @@ impl Endpoint {
...
+                    ..
+                })),
+                Some(FnArg::Typed(PatType { ty, pat, .. })),
+            ) => {
+                sender = Some((id_pat.clone(), id_ty.clone()));
                 arg = Some((pat.clone(), ty.clone()));
             }
             (Some(FnArg::Typed(PatType { ty, pat, .. })), None) => {
+                sender = None;
                 arg = Some((pat.clone(), ty.clone()));
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-macros/src/lib.rs:134]: If there's only 1 argument, we assume it's the argument. The local variables are kinda wrong in this case, it's more like `maybe_arg1, maybe_arg2`.

File: src/many-identity/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+cryptoki = { version = "0.3", features = ["generate-bindings"] }
+coset = "0.3.2"
+ed25519 = { version = "1.2.0", features = [ "std" ] }
+ed25519-dalek = { version = "1.0.1", features = [ "std" ] }
+hex = "0.4.3"
+minicbor = "0.18.0"
+once_cell = "1.10"
+p256 = { version = "0.9.0", features = [ "pem", "ecdsa", "std" ] }
+pkcs8 = { version = "0.8", features = [ "pem", "std", "encryption" ] }
+rand = "0.7.3"
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-identity/Cargo.toml:22]: Ah I was wondering about that.

File: src/many-types/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+many-error = { path = "../many-error", version = "0.1.0" }
+many-identity = { path = "../many-identity", version = "0.1.0" }
+coset = "0.3.2"
+fixed = "1.16.0"
+hex = "0.4.3"
+minicbor = { version = "0.18.0", features = ["derive", "std", "half"] }
+num-derive = "0.3.3"
+num-traits = "0.2.14"
+num-bigint = "0.4.3"
+proptest = "1.0.0"
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-types/Cargo.toml:18]: No (because we export testing stuff that uses it), but it should be optional. Added a feature.

File: src/many-types/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+many-error = { path = "../many-error", version = "0.1.0" }
+many-identity = { path = "../many-identity", version = "0.1.0" }
+coset = "0.3.2"
+fixed = "1.16.0"
+hex = "0.4.3"
+minicbor = { version = "0.18.0", features = ["derive", "std", "half"] }
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-types/Cargo.toml:14]: Lots of work and edge cases, probably slipped under my radar. Done.

File: src/many-types/src/cbor.rs [hunk 10]
```diff
@@ -90,7 +90,6 @@ impl<'d, C> Decode<'d, C> for CborAny {
     }
 }
 
-#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-types/src/cbor.rs:93]: `arb_cbor` used to be only used by the `many` crate, now it's needed by more. I hid it behind the `proptest` feature.

File: src/many-server/src/message/request.rs [hunk 11]
```diff
@@ -1,6 +1,6 @@
 use crate::protocol::attributes::AttributeSet;
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-server/src/message/request.rs:1]: Yes, removed. Good catch!

File: src/many-protocol/src/attributes.rs [hunk 12]
```diff
@@ -1,14 +1,69 @@
-use crate::cbor::CborAny;
+use many_error::ManyError;
+use many_types::cbor::CborAny;
+use many_types::AttributeId;
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-protocol/src/attributes.rs:3]: Moved to `types`. Agreed

File: src/many/src/main.rs (hunk 1 above)
Comment (hansl) [reply src/many/src/main.rs:163]: Refactoring in IntelliJ isn't _perfect_.

**Comment by hansl** [comment]
On Monday will do the diagram and all.

**Review by fmorency — APPROVED** [review]
LGTM, assuming the CI is fixed.

I hate that I had to do that for CI but there's no way to do something like `--all-features --disable-feature foo`

Thanks!

**Comment by fmorency** [comment]
And thanks for putting those on crates.io! 👍🏻

**Review by fmorency — APPROVED** [review]
